apps/pkeyutl.c

```diff
@@ -40,6 +43,14 @@ static int do_raw_keyop(int pkey_op, EVP_MD_CTX *mctx,
                         int filesize, unsigned char *sig, int siglen,
                         unsigned char **out, size_t *poutlen);
 
+static int only_rawin(const EVP_PKEY *pkey)
+{
+    if (pkey == NULL)
+        return 0;
+    return EVP_PKEY_is_a(pkey, "ED25519")
+        || EVP_PKEY_is_a(pkey, "ED448");
```

Comment:
Should anything be done to ensure that ECDSA signatures are with a digest of ideally the same (or perhaps shorter) length, since the underlying ECDSA signature scheme truncates inputs.

Indeed, one would hope the ECDSA signatures are always with the *default* digest for the algorithm, making `-rawin` the preferred way for pretty much all operations, feeding in already digested data is most unusual, and requires care to get it right.

Reply:
Very good point!!
I found that the documentation falsely claims on the `-digest` option:

> If this option is omitted but the signature algorithm requires one, a default value will be
> used. For signature algorithms like RSA, DSA and ECDSA, SHA-256 will be the default digest algorithm.
> For SM2, it will be SM3.

I'm going to file a bug report on that later today...

For now, I've just strengthened the existing error message in case the input is more than 512 bits long
without a digest given, which is a clear indication that the user failed to provide it.

I'd like to add a warning (to allow for backward compatibility and for special cases where this may be intended)
in case the input is longer than the input size of the asymmetric signing operation.
To this end: is there an API function for getting that value for any given EVP_PKEY?

Reply:
> Very good point!! I found that the documentation falsely claims on the `-digest` option:

The claim does not look false to me, whether signing and verification work equally well with an implicit digest.
```
$ openssl pkeyutl -in msg -sign -inkey p256.pem -rawin -out p256-sig-default.dat
$ openssl pkeyutl -in msg -sign -inkey p256.pem -rawin -digest sha256 -out p256-sig-sha256.dat

$ openssl pkeyutl -rawin -in msg -sigfile p256-sig-default.dat -verify -inkey p256.pem
Signature Verified Successfully

$ openssl pkeyutl -rawin -in msg -sigfile p256-sig-default.dat -digest sha256 -verify -inkey p256.pem
Signature Verified Successfully

$ openssl pkeyutl -rawin -in msg -sigfile p256-sig-sha256.dat -digest sha256 -verify -inkey p256.pem
Signature Verified Successfully

$ openssl pkeyutl -rawin -in msg -sigfile p256-sig-sha256.dat -verify -inkey p256.pem
Signature Verified Successfully
```

Reply:
You did not say which size `msg` has in your case,
but likely this is immaterial for your above tests because you use `-rawin` in all of them.
Yet the bug, as reported in #25827, is on the case that `-rawin` is not given.

Note that the above quoted documentation of `-digest` does not include the precondition that `-rawin` is given.
Maybe the issue is simply that this was forgotten to mention?

Reply:
Actually, the `-digest` document does say that it's only valid together with `-rawin`.  Quoting just part of the `-digest` documentation hides this detail.

Reply:
So how exactly is one supposed to sign pre-digested input with RSA PKCS#1? Without `-digest` it seems that the `rawin` input needs to include the PKCS#1 digest OID and parameters, as well as ASN.1 encoding of the digest, otherwise I get:

```
$ openssl pkeyutl -in msghash.dat -sign -inkey rsa.pem -out rsa-sig-default.dat

$ openssl pkeyutl -inkey rsa.pem -encrypt -pkeyopt rsa_padding_mode:none -in rsa-sig-default.dat | od -tx1
0000000 00 01 ff ff ff ff ff ff ff ff ff ff ff ff ff ff
0000020 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
*
0000360 00 41 74 74 61 63 6b 20 61 74 20 64 61 77 6e 0a
```
In which the PKCS#1 payload is just the hash, which isn't that useful.

Reply:
> Actually, the `-digest` document does say that it's only valid together with `-rawin`. 
> Quoting just part of the `-digest` documentation hides this detail.

Not really - the doc says there:
> If this option is present, then the `-rawin` option must be also specified.

So -digest cannot be used without -rawin, but when -digest is omitted, nothing is implied for -rawin.
The -rawin option is pretty confusing, and the aim of this PR to make it more or less implicit.

Reply:
The `-rawin` option is pretty confusing, and the aim of this PR is, for usability. to make it more or less implicit.

Reply:
Though, admittedly in the case of ECDSA (IIRC) the digest OID is not part of the signed data, one just signs the digest value, and it is up to verifyer to know the digest in advance. So the need to encode a digest OID in the signature input looks RSA-specific.

Reply:
Actually, what you're doing there is documented in the NOTES about `rsa_padding_mode:{mode}`:

https://github.com/openssl/openssl/blob/a3660729e68dc11c01edb4a349ff2610b6b59ee0/doc/man1/openssl-pkeyutl.pod.in#L273-L276

The "secret" for setting the digest in this case is another `-pkeyopt`, see the beginning of NOTES:

https://github.com/openssl/openssl/blob/a3660729e68dc11c01edb4a349ff2610b6b59ee0/doc/man1/openssl-pkeyutl.pod.in#L238-L239

So your experiment gave the documented result, and the way to get a signature containing a full DigestInfo looks like this:

``` console
$ openssl pkeyutl -in <(echo whatever | openssl sha256 -binary) -sign -inkey rsa.pem -pkeyopt digest:sha256 -out rsa-sig-default.dat
$ openssl pkeyutl -inkey rsa.pem -encrypt -pkeyopt rsa_padding_mode:none -in rsa-sig-default.dat | od -tx1
0000000 00 01 ff ff ff ff ff ff ff ff ff ff ff ff ff ff
0000020 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
*
0000300 ff ff ff ff ff ff ff ff ff ff ff ff 00 30 31 30
0000320 0d 06 09 60 86 48 01 65 03 04 02 01 05 00 04 20
0000340 cd 29 3b e6 ce a0 34 bd 45 a0 35 27 75 a2 19 ef
0000360 5d c7 82 5c e5 5d 1f 7d ae 97 62 d8 0c e6 44 11
0000400
```

Reply:
> So the need to encode a digest OID in the signature input looks RSA-specific.

Yup...  or to be more precise, it is specific for PKCS#1.  That's why you have to give it as a `-pkeyopt`, 'cause it's only applicable for some algorithms, not others.